[pylint] Implement `consider-using-assignment-expr` (`R6103`)

vincevannoort opened the pull request:
## Summary

This pull request implements the `R6103` pylint rule: [pylint documentation](https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/consider-using-assignment-expr.html).

It checks assignments which are directly followed by if statements using that expression:

```python
test1 = "example"
if test1:
    print(test1)
```

And suggest to use the `:=` operator to 

```python
if test1 := "example":
    print(test1)
```

## Test Plan

I have added test cases, and checked some of the `ruff ecosystem` results.

vincevannoort commented on crates/ruff_linter/src/checkers/ast/analyze/statement.rs:
```diff
@@ -1131,6 +1131,9 @@ pub(crate) fn statement(stmt: &Stmt, checker: &mut Checker) {
             if checker.enabled(Rule::TooManyNestedBlocks) {
                 pylint::rules::too_many_nested_blocks(checker, stmt);
             }
+            if checker.enabled(Rule::UnnecessaryAssignment) {
+                pylint::rules::unnecessary_assignment(checker, if_);
+            }
```
I applied this rule by checking whether the previous statement from a `IfStmt` is an `AssignStmt`, however I am wondering if it is more desirable to do check if the next statement of a `AssignStmt` is an `IfStmt`?

vincevannoort replied:
I agree that assignments are far more common than if statements.

I think the answer depends on the costs for checking `previous_statement` and `next_statement`.

My thought here is that retrieving the previous statement using the newly added `previous_statement` might be more expensive because it has to iterate over all previous statements using `previous_statements` to find the previous statement (if there is a better way, please let me know).

While checking an assignment, then checking the `next_statement` might be a cheap operation.

Do you have any idea? If they have equal cost I think the current implementation is fine. 😄

vincevannoort replied:
I can change it, to the suggested approach if you want 😄

vincevannoort commented on crates/ruff_python_semantic/src/model.rs:
```diff
@@ -1247,6 +1247,28 @@ impl<'a> SemanticModel<'a> {
             .nth(1)
     }
 
+    pub fn previous_statement(&self, stmt: &'a Stmt) -> Option<&Stmt> {
+        self.previous_statements(stmt)?.next()
+    }
```
I needed something like this function, not sure if there is an existing better way?

MichaReiser replied:
That makes sense. I think we can make it more efficient with https://github.com/astral-sh/ruff/pull/13895
which reduces an extra find and collect

sbrugman commented on crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs:
````diff
@@ -0,0 +1,199 @@
+/// ```python
+/// if test1 := "example":
+///     print(test1)
+/// ```
````
The Python 3.8 release notes would also make a great resource:

https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions

sbrugman commented on crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs:
````diff
@@ -0,0 +1,199 @@
+/// ```python
+/// if test1 := "example":
+///     print(test1)
+/// ```
````
```suggestion
/// ```

## References

- [Python Documentation: What’s New In Python 3.8](https://docs.python.org/3/whatsnew/3.8.html#assignment-expressions)
- [PEP 572 – Assignment Expressions](https://peps.python.org/pep-0572/)
```

vincevannoort replied:
Good suggestion! 😄

MichaReiser reviewed (changes requested):
Thanks for this contribution. My plane is about to land. I've to finish the review at a later time. 

I only had enough time to quickly glance over the Rust code. We should look into removing the many `node.clone()` calls because it is fairly expensive to clone nodes and probably unnecessary. Let me know if you need some guidance on how to remove the clone calls (It probably requires adding some lifetimes)

Regarding the rule's naming. 

* I did a quick search to see how we referred to `:=` in other rules. There are not many usages but `named expression (walrus operator)` is the most common form.
* The rule name seems too generic to me and its name is very similar to [`unnecessary-assign`](https://docs.astral.sh/ruff/rules/unnecessary-assign/) (which we should rename to `unnecessary-assignment). Reading through the examples the rule mainly is about assigning a value that is then only used in an if condition. I need to think a bit more about what a good rule name could be. Maybe you have an idea?

vincevannoort commented:
> Thanks for this contribution. My plane is about to land. I've to finish the review at a later time.

Thanks for the review, much appreciated! 👍 

> I only had enough time to quickly glance over the Rust code. We should look into removing the many `node.clone()` calls because it is fairly expensive to clone nodes and probably unnecessary. Let me know if you need some guidance on how to remove the clone calls (It probably requires adding some lifetimes)

I have tried and was able to remove almost all clone calls, except a few which I think are needed for returning the diagnostic. Could you take a look at the remaining ones and see if any of the 3 can still be removed?

> Regarding the rule's naming.
> 
> * I did a quick search to see how we referred to `:=` in other rules. There are not many usages but `named expression (walrus operator)` is the most common form.
> * The rule name seems too generic to me and its name is very similar to [`unnecessary-assign`](https://docs.astral.sh/ruff/rules/unnecessary-assign/) (which we should rename to `unnecessary-assignment). Reading through the examples the rule mainly is about assigning a value that is then only used in an if condition. I need to think a bit more about what a good rule name could be. Maybe you have an idea?

I agree, here are some possible options:

1. unnecessary_assignment_before_if_stmt
2. redundant_assignment_before_if _stmt
3. standalone_assignment_before_if_stmt

These seem close to what the lint is trying to prevent, do you have other ideas in mind?

vincevannoort commented on crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs:
```diff
@@ -0,0 +1,203 @@
+        return;
+    }
+
+    let if_test = *stmt.test.clone();
```
I have no motivation other than that I had trouble with not cloning these before, but I have been able to make it work.

vincevannoort commented on crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_assignment.py:
```diff
@@ -0,0 +1,70 @@
+    if good5_5: # correct, assignment is not the previous statement
+        pass
+
+for good6 in [1, 2, 3]:
```
I have added some examples for this now!

vincevannoort commented:
Hey @MichaReiser, once you have the time, would you mind giving this pull request another look? :smile:

vincevannoort commented:
Hey @MichaReiser, could you or someone else from the team have a look? :smile:

MichaReiser commented on crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs:
```diff
@@ -0,0 +1,202 @@
+        };
+    }
+
+    // case - elif else clauses (`elif test1:`)
```
Is there a reason why compare expressions aren't handled inside `elif_else` clauses?

MichaReiser commented on crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs:
```diff
@@ -0,0 +1,202 @@
+    }
+
+    // case - elif else clauses (`elif test1:`)
+    errors.extend(
+        stmt.elif_else_clauses
+            .iter()
+            .filter(|elif_else_clause| elif_else_clause.test.is_some())
+            .filter_map(|elif_else_clause| {
+                let elif_check = elif_else_clause.test.as_ref().unwrap();
+                find_assignment_before_if_stmt(semantic, elif_check, elif_check)
+            })
+            .collect::<Vec<AssignmentBeforeIfStmt>>(),
```
Changing assignments to walrus operators in `elif else` branches is semantically incorrect if the variable is used afterwards

```python
>>> if True: ...
... elif x :=10: ...
... 
Ellipsis
>>> x
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'x' is not defined
```

vincevannoort replied:
Yeah, I think you are right. Due to that I decided to only apply the rule to simple `if` statements, where simple means without `elif` or `else` attached to the `if`. That should prevent cases like this.

MichaReiser commented on crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs:
```diff
@@ -0,0 +1,202 @@
+    error: AssignmentBeforeIfStmt,
+) -> Diagnostic {
+    let (origin, expr_name, assignment) = error;
+    let assignment_expr = generator.expr(&assignment.value);
```
What's the motivation for calling into the generator here? We try to avoid using the generator because it removes comments. Could we instead take the assignment value as it is in the source (using locator)? Note: We have to be careful about parenthesized expressions.

MichaReiser commented on crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_assignment.py:
```diff
@@ -0,0 +1,78 @@
+if bad6 is not None: # [consider-using-assignment-expr]
+    pass
+
+bad7 = 'example'
+if bad7 == 'something': # [consider-using-assignment-expr]
+    pass
+elif bad7 == 'something else':
+    pass
```
This is an interesting example and possibly controversial. I would prefer the existing solution because the assignment in the `if` is very subtle and I can see how it can be confusing when trying to figure out what the value of `bad7` is in the `elif` branch

vincevannoort replied:
Yeah, I think you are right. Due to that I decided to only apply the rule to simple if statements, where simple means without `elif` or `else` attached to the `if`.

MichaReiser reviewed (changes requested):
Thanks for the ping and sorry for the late review. 

I created a PR that should allow us to implement a more efficient `previous_statement`  here. 

I think the rule has to become cleverer, at least when we want to support handling `elif` cases because today's implementation can result in changes that fail at runtime.

vincevannoort commented:
Thanks for the review @MichaReiser :+1: , just a heads up: I will be travelling without my laptop for the coming 5 weeks, so will only get back to pull request this after.

vincevannoort commented on crates/ruff_linter/resources/test/fixtures/pylint/unnecessary_assignment.py:
```diff
@@ -0,0 +1,70 @@
+    if bad4: # [consider-using-assignment-expr]
+        pass
+
+bad5 = (
+    'example',
+    'example',
+    'example',
+    'example',
+    'example',
+    'example',
+    'example',
+    'example',
+    'example',
+    'example',
+)
```
I agree, should we check the complexity of the assignment to determine whether to apply the rule? If so, how do we determine where to draw the line?

vincevannoort commented:
@MichaReiser sorry for getting back so late to this PR, and thanks for all the preivous input.

I have rebased to the latest changes on main, and fixed some issues mentioned in earlier comments. Additionally, I changed the rule to only apply to simple `if` statements, that do not have an `elif` or `else` attached. This should hopefully resolve most of the more controversial cases.

Coul you take another look?

vincevannoort commented on crates/ruff_linter/src/rules/pylint/rules/unnecessary_assignment.rs:
```diff
@@ -0,0 +1,204 @@
+        diagnostics.push(create_diagnostic(checker, unreferenced_binding));
+    };
+
+    match &*stmt.test {
+        // case - bool operations (`if test1 and test2:`)
+        Expr::BoolOp(expr) => diagnostics.extend(expr.values.iter().filter_map(|bool_test| {
+            Some(create_diagnostic(
+                checker,
+                find_assignment_before_if_stmt(&previous_assignment, if_test, bool_test)?,
+            ))
+        })),
+
+        // case - compare (`if test1 is not None:`)
+        Expr::Compare(compare) => {
+            if let Some(error) =
+                find_assignment_before_if_stmt(&previous_assignment, if_test, &compare.left)
+            {
+                diagnostics.push(create_diagnostic(checker, error));
+            };
+        }
+
+        _ => {}
+    }
```
I am curious if we should actually do this part. If I look at the ecosystem checks, these also produce some possible controversial results. 🤔 

Without it, I think most of the more controversial results would disappear.

MichaReiser commented:
Hmm, I just noticed that this is a pylint extensions rule which we don't want to add right now because it would require a new rule group.

vincevannoort commented:
> Hmm, I just noticed that this is a pylint extensions rule which we don't want to add right now because it would require a new rule group.

What would you recommend in this case? Just close the pull request or make it a new ruff rule instead of pylint rule?

MichaReiser commented:
Yeah, I think we have to close this for now because there's no clear path to landing this PR before #1774 is complete. I'm very sorry for wasting your time here. I should have noticed this way sooner.

vincevannoort commented:
> Yeah, I think we have to close this for now because there's no clear path to landing this PR before #1774 is complete. I'm very sorry for wasting your time here. I should have noticed this way sooner. But thank you anyway and we can come back to this PR once we're at a point where we accept pylint extension rules.

No worries :smile: , I still learned quite a lot from the implementation so it was still worth it for me personally. Thanks for the effort in the feedback rounds. :+1:
